lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -377,21 +480,223 @@ instance ToText tr => ToText (FollowLog tr) where
+        s { tip = overCurrent (const $ slotFromMaybeBh p) (tip s) }
+    _ -> s
+  where
+    slotFromMaybeBh = maybe (SlotNo 0) slotNo
+
+instance ToText (FollowStats LogState) where
+    toText st@(FollowStats b r tip t prog) = syncStatus <> " " <> stats <> sevExpl
+      where
+        syncStatus = case prog of
+            LogState NotResponding Ready ->
+                "In sync."
+            LogState Ready Ready ->
+                "Still in sync."
+            LogState NotResponding NotResponding ->
+                "Still not syncing."
+            LogState (Syncing _p) Ready ->
+                "In sync!"
+            LogState Ready (Syncing p) ->
+                "Fell out of sync (" <> (pretty p) <> ")"
+            LogState _ (Syncing p) ->
+                "Syncing (" <> (pretty p) <> ")"
+            LogState prev NotResponding ->
+                "Not responding. Previously " <> (pretty prev) <> "."
+        stats = mconcat
+            [ "Applied " <> pretty (using (-) b) <> " blocks, "
+            , pretty (using (-) r) <> " rollbacks "
+            , "in the last " <> pretty (using diffUTCTime t) <> ". "
+            , "Currently at slot " <> pretty (current tip) <> "."
+            ]
+          where
+            using f x = f (current x) (prev x)
+
+        sevExpl = maybe
+            ""
+            (\x -> " (" <> x <> ")")
+            (snd $ explainedSeverityAnnotation st)
+
+-- NOTE: Here we check if the sync progress is going backwards, which
+-- would be a sign the wallet is overloaded (or rollbacks)
+--
+-- But this check might be in the wrong place. Might be better to
+-- produce new logs from inside the updateStats function and immeditely
+-- warn there.
+explainedSeverityAnnotation :: FollowStats LogState -> (Severity, Maybe Text)
+explainedSeverityAnnotation s
+        | progressMovedBackwards = (Warning, Just "progress decreased")
+        | noBlocks && notRestored = (Warning, Just "not applying blocks")
+        | nowInSync = (Notice, Nothing)
+        | otherwise = (Info, Nothing)
+  where
+    progressMovedBackwards = current (prog s) < prev (prog s)
+    nowInSync = current (prog s) == Ready && prev (prog s) < Ready
+    notRestored = current (prog s) /= Ready
+    noBlocks = (current (blocksApplied s) - prev (blocksApplied s)) <= 0
+
+
+instance HasSeverityAnnotation (FollowStats LogState) where
+    getSeverityAnnotation = fst . explainedSeverityAnnotation
+
+-- | Starts a new thread for monitoring health and statistics from
+-- the returned @FollowLog tr@.
+withFollowStatsMonitoring
+    :: Tracer IO (FollowLog tr)
+    -> (SlotNo -> IO SyncProgress)
+    -> ((Tracer IO (FollowLog tr)) -> IO ())
+    -> IO ()
+withFollowStatsMonitoring tr calcSyncProgress act = do
+    t0' <- getCurrentTime
+    var <- newTMVarIO $ emptyStats t0'
+    let tr' = flip contramapM tr $ \msg -> do
+            atomically $ do
+                s <- takeTMVar var
+                putTMVar var (updateStats msg s)
```

Comment:
This is not going to leave unevaluated thunks, is it? I suppose all the results are consumed by the other thread.

Reply:
Good that you brought it up again. I did add a `force` in `updateStats` long time ago, but:
1. I'm not sure if that was effective or not because the outer lazy `putTMVar`
2. After asking in #haskell it seems it seems `force` (but ok, it also sounds familiar) is not the right way

Reply:
I replaced `TMVar` with `StrictTMVar` for now (could have used `$!` on this line also I think).

Will have a look at how to do it better in a follow-up PR.